enterprise/internal/scim/user_delete.go

```diff
@@ -0,0 +1,107 @@
+package scim
+
+import (
+	"context"
+	"net/http"
+	"strconv"
+
+	"github.com/sourcegraph/log"
+	"github.com/sourcegraph/sourcegraph/internal/authz"
+	"github.com/sourcegraph/sourcegraph/internal/database"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc"
+	"github.com/sourcegraph/sourcegraph/internal/types"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// Delete removes the resource with corresponding ID.
+func (h *UserResourceHandler) Delete(r *http.Request, id string) error {
+	logger := h.observationCtx.Logger.Scoped("DeleteUsers", "SCIM delete user").With(log.String("user", id))
+
+	// Find user
+	idInt, err := strconv.Atoi(id)
+	if err != nil {
+		return errors.Wrap(err, "parse user ID")
+	}
+	user, err := findUser(r.Context(), h.db, idInt, logger)
+	if err != nil {
+		return err
+	}
+
+	// If we found no user, we report “all clear” to match the spec
+	if user.Username == "" {
+		return nil
+	}
+
+	// Save username, verified emails, and external accounts to be used for revoking user permissions after deletion
+	revokeUserPermissionsArgsList, err := getRevokeUserPermissionArgs(r.Context(), user, h.db)
+	if err != nil {
+		return err
+	}
+
+	// Delete user
+	if err := h.db.Users().HardDelete(r.Context(), int32(idInt)); err != nil {
+		return err
+	}
+
+	// NOTE: Practically, we don't reuse the ID for any new users, and the situation of left-over pending permissions
+	// is possible but highly unlikely. Therefore, there is no need to roll back user deletion even if this step failed.
+	// This call is purely for the purpose of cleanup.
+	if err := h.db.Authz().RevokeUserPermissionsList(r.Context(), []*database.RevokeUserPermissionsArgs{revokeUserPermissionsArgsList}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// findUser finds the user with the given ID. If the user does not exist, it returns an empty user.
+func findUser(ctx context.Context, db database.DB, id int, logger log.Logger) (types.UserForSCIM, error) {
+	users, err := db.Users().ListForSCIM(ctx, &database.UsersListOptions{
+		UserIDs: []int32{int32(id)},
+	})
+	if err != nil {
+		return types.UserForSCIM{}, errors.Wrap(err, "list users by IDs")
+	}
+	if len(users) == 0 {
+		logger.Info("requested users to delete do not exist")
+		return types.UserForSCIM{}, nil
+	} else if len(users) > 1 {
+		logger.Error("requested users to delete have duplicate IDs—that should not happen")
+		return types.UserForSCIM{}, errors.New("requested users to delete have duplicate IDs")
+	} else {
+		logger.Debug("attempting to delete requested users")
+	}
```

Comment:
I think we should change the messaging here. The method is called `findUser` and that should be it's responsibility. We do not care if the caller is using it to delete the user or do something completely different. We should also scope the logger for this function IMO.
```suggestion
	if len(users) == 0 {
		logger.Info("no match found for", log.Int("userID", id"))
		return types.UserForSCIM{}, nil
	} else if len(users) > 1 {
		logger.Error("duplicate IDs found - that should not happen")
		return types.UserForSCIM{}, errors.New("multiple users match the find criteria")
	} else {
		logger.Debug("findUsers found exactly 1 result")
	}
```

Reply:
Makes sense, will change this.

Reply:
On second thought, I removed the logger altogether. We get all the info in the returned errors, so it didn't add real value.